### DISCUSSION TITLE
Why git-cliff must be executed from root directory?

### DISCUSSION THREAD

**Discussion by azzamsa** [orhun/git-cliff#426] [Q&A] [answered]
Hi, 👋.

Why is it that, unlike `git`, `git-cliff` can't be executed from a sub-directory?
`git` can be run from any child directory without being executed where the `.git` directory is located.

```bash
nvim/configs on  master [!⇕]
🦄 git-cliff --tag v2.2.0 --output CHANGELOG.md --workdir .
 ERROR git_cliff > Git error: `could not find the repository at '.'; class=Repository (6); code=NotFound (-3)`

nvim/configs on  master [!⇕]
🙊 git-cliff --tag v2.2.0 --output CHANGELOG.md --workdir ..
 WARN  git_cliff > "../cliff.toml" is not found, using the default configuration.
```

The first command fails because it runs in `nvim/configs`, but the second succeeds because I passed `--workdir ..`.
However, when we observe git behavior:

```bash
nvim/configs on  master [!?⇕]
🦄 git status
On branch master
Your branch and 'origin/master' have diverged,
```

It works from any location.

---

Why must it be executed from the root directory?
If you're okay with it, can I make a PR where git-cliff searches for `.git` in the parent directory, then executes the operations if found; otherwise, it gives up.

The implementation would be similar to https://github.com/azzamsa/toor/blob/579ecc9a113a2a66dd5f46a4986786251420ec1b/src/project.rs#L20

**Comment by alerque** [comment]
This should be easily fixed, but don't do it by searching for `.git`. It is possible that the directory has a different name or even is located outside of the tree altogether. Also that would not play nice with submodules. Instead just use the appropriate API in libgit (git2 crate) to find the repository root. That will account for a lot of edge cases you might otherwise get wrong.

**Comment by orhun** [comment]
Hello!

> If you're okay with it, can I make a PR where git-cliff searches for .git in the parent directory, then executes the operations if found; otherwise, it gives up.

It makes sense to have this feature so please go ahead!

As @alerque mentioned, this should be handled with `git2`.

**Comment by RJSonnenberg** [comment]
@azzamsa Any movement on this? Just curious if any progress has been made here.

**Reply by alerque** [reply]
This would be a relatively easy thing to contribute yourself.

**Reply by orhun** [reply]
Yup, please go ahead @RJSonnenberg if you are interested :)

**Reply by RJSonnenberg** [reply]
I would love to, but I'm at a complete loss of where to start/look. I am not familiar with Rust or Rust-based project structure. Any hints on where to start?

**Reply by orhun** [reply]
I think you can take a look at `Repository::init` function in `repo.rs` - this is where the magic should happen :)

You can use [`GitRepository::discover`](https://docs.rs/git2/latest/git2/struct.Repository.html#method.discover) function instead of `open`, and I guess that should be it!

**Comment by orhun** [comment, marked as answer]
Implemented in #975

**Reply by azzamsa** [reply]
Finally, thanks! 🍩
